rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,144 @@
+
+    pub fn name(&self) -> String;
+
+    pub fn merge(
+        &self,
+        key: Bytes,
+        existing_value: Bytes,
+        value: Bytes
+    ) -> Result<Bytes, MergeOperatorError>;
+}
+```
+
+Initially, there will be a single optional instance of the merge operator per database. The user needs to ensure that the compactor and writer run with the same merge operator to get predictable results.
+
+## Implementation
+
+### Data Structures
+
+We need to introduce an additional variant of `ValueDeletable` that denotes a **Merge Operand**. As a result of this change, we can no longer represent the deletable as `Option<Bytes>` or `Option<&[u8]>`, because a non-empty state can represent either a value or a merge operand. This requires reusing the ValueDeletable data structure throughout the stack—all the way from the memtable to the row codec.
+
+```rust
+#[derive(Debug, Clone, PartialEq)]
+pub enum ValueDeletable {
+    ...
+    Merge(Bytes)
+    ...
+}
+```
+
+### Row Codec
+
+We need to slightly adjust the binary format of the SST file and introduce a new flag to denote a merge operand. Note that the `Tombstone` and `Merge` flags are mutually exclusive.
+
+Aside from this, `Merge` rows can follow the same structure as `Value` rows.
+
+This change is backwards compatible.
+
+### Read Path
+
+Merges are expected to be read using the standard `get` API.
+
+For each key, SlateDB maintains a history of operations, which is subject to compaction.
+
+Let's denote each operation as `OPi`. A key (`K`) that has experienced `n` changes looks like this logically (physically, the changes could be spread across the memtable and multiple SST files):
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+
+earliest --------------------> latest
+```
+
+Currently, during a read operation, SlateDB always observes the latest known state of the database. In other words, it only needs to look up the latest operation. If it's a `Value`, it simply returns it, and if it's a `Tombstone`, it returns `None`. All previous operations can be ignored and will be removed by compaction at some point.
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+                                   ^
+                                   |
+                                  GET
+```
+
+With the addition of a **Merge Operand**, it may be necessary to look backwards at previous values, up to the point where we encounter either a `Value` or a `Tombstone`. Everything beyond that point can be ignored.
+
+```
+K:    OP1    OP2    OP3    OP4    OPn
+            Value  Merge  Merge  Merge
+                                   ^
+                                   |
+                                  GET
+```
+
+
+In the above example, `get` should essentially return something like:
+
+```
+Merge(OP2, Merge(OP3, Merge(OP4, OPn)))
+```
+
+### Compactions
+
+Compactions will continuously reduce the history by using the same mechanics as described above, performing partial merges along the way.
```

Comment:
Snapshot isolation is going to make this a bit tricky. We can't apply a merge between rows with two sequence numbers N and N' if there is a snapshot open that references a sequence number M is between N and N'. We don't support this yet, but I think @pragmaticanon is working on an RFC so we should make sure this doesn't get lost. It probably makes sense to address in that RFC since there's a similar problem for just general compaction of rows.

Reply:
Tagging #260 to link this discussion.

Reply:
I'll add a note on this. RocksDB seems to already solve this by only merging up to the snapshot barrier.

Reply:
added